rdtools/system_analysis.py

```diff
@@ -0,0 +1,672 @@
+
+        loc = self.pvlib_location
+        sun = loc.get_solarposition(times)
+        clearsky = loc.get_clearsky(times, solar_position=sun)
+
+        clearsky_poa = pvlib.irradiance.get_total_irradiance(self.pv_tilt, self.pv_azimuth, sun['apparent_zenith'],
+                                                             sun['azimuth'], clearsky['dni'], clearsky['ghi'],
+                                                             clearsky['dhi'], albedo=self.albedo, **kwargs)
+        clearsky_poa = clearsky_poa['poa_global']
+
+        if rescale is True:
+            clearsky_poa = normalization.irradiance_rescale(self.poa, clearsky_poa, method='iterative')
+
+        self.clearsky_poa = clearsky_poa
+
+    def calc_cell_temperature(self, poa, windspeed, ambient_temperature):
+        '''
+        Return cell temperature calculated from ambient conditions.
+
+        Parameters
+        ----------
+        poa : numeric
+            Plane of array irradiance in W/m^2
+        windspeed = numeric
+            Wind speed in m/s
+        ambient_temperature : numeric
+            Ambient temperature in Celsius
+
+        Returns
+        -------
+        numeric
+            calculated cell temperature
+        '''
+        if self.temperature_model is None:
+            cell_temp = pvlib.pvsystem.sapm_celltemp(poa, windspeed, ambient_temperature)
+        else:
+            cell_temp = pvlib.pvsystem.sapm_celltemp(poa, windspeed, ambient_temperature, model=self.temperature_model)
+        cell_temp = cell_temp['temp_cell']
+
+        return cell_temp
+
+    def calc_clearsky_tamb(self):
+        '''
+        Calculate clear-sky ambient temperature and store in self.clearsky_ambient_temperature
+        '''
+        times = self.clearsky_poa.index
+        if self.pvlib_location is None:
+            raise ValueError('pvlib location must be provided')
+        loc = self.pvlib_location
+
+        cs_amb_temp = clearsky_temperature.get_clearsky_tamb(times, loc.latitude, loc.longitude)
+
+        self.clearsky_ambient_temperature = cs_amb_temp
+
+    def pvwatts_norm(self, poa, cell_temperature):
+        '''
+        Normalize PV energy to that expected from a PVWatts model.
+
+        Parameters
+        ---------
+        poa : numeric
+            plane of array irradiance in W/m^2
+        cell_temperature : numeric
+            cell temperature in Celsius
+
+        Returns
+        -------
+        pandas.Series
+            Normalized pv energy
+        pandas.Series
+            Associated insolation
+        '''
+        if self.pv_nameplate is None:
+            renorm = True
+            pv_nameplate = 1.0
+        else:
+            renorm = False
+
+        if self.temperature_coefficient is None:
+            raise ValueError('Temperature coeffcient must be available to perform pvwatts_norm')
+
+        pvwatts_kws = {"poa_global": poa,
+                       "P_ref": pv_nameplate,
```

Comment:
`pv_nameplate` is only defined if `self.pv_nameplate is None`, this throws an UnboundLocalVariable error if `pv_nameplate` is given to the constructor.

Reply:
Fixed by #133